src/algebraic_geometry/locally_ringed_space.lean

```diff
@@ -0,0 +1,123 @@
+such that all the stalks are local rings.
+
+A morphism of locally ringed spaces is a morphism of ringed spaces
+such that the morphims induced on stalks are local ring homomorphisms. -/
+@[nolint has_inhabited_instance]
+structure LocallyRingedSpace extends SheafedSpace CommRing :=
+(local_ring : ∀ x, local_ring (presheaf.stalk x))
+
+attribute [instance] LocallyRingedSpace.local_ring
+
+namespace LocallyRingedSpace
+
+variables (X : LocallyRingedSpace)
+
+/-- The underlying topological space of a locally ringed space. -/
+def to_Top : Top := X.1.carrier
+
+instance : has_coe_to_sort LocallyRingedSpace :=
+{ S := Type u,
+  coe := λ X : LocallyRingedSpace, (X.to_Top : Type u), }
+
+-- PROJECT: how about a typeclass "has_structure_sheaf" to mediate the 𝒪 notation, rather
+-- than defining it over and over for PresheafedSpace, LRS, Scheme, etc.
+
+/-- The structure sheaf of a locally ringed space. -/
+def 𝒪 : sheaf CommRing X.to_Top := ⟨X.1.presheaf, X.1.sheaf_condition⟩
+
+/-- A morphism of locally ringed spaces is a morphism of ringed spaces
+ such that the morphims induced on stalks are local ring homomorphisms. -/
+def hom (X Y : LocallyRingedSpace) : Type* :=
+{ f : X.to_SheafedSpace ⟶ Y.to_SheafedSpace // ∀ x, is_local_ring_hom (PresheafedSpace.stalk_map f x) }
+
+instance : has_hom LocallyRingedSpace := ⟨hom⟩
+
+@[ext] lemma hom_ext {X Y : LocallyRingedSpace} (f g : hom X Y) (w : f.1 = g.1) : f = g :=
+subtype.eq w
+
+/--
+The stalk of a locally ringed space, just as a `CommRing`.
+-/
+-- TODO perhaps we should make a bundled `LocalRing` and return one here?
+-- TODO define `sheaf.stalk` so we can write `X.𝒪.stalk` here?
+def stalk (X : LocallyRingedSpace) (x : X) := X.presheaf.stalk x
+
+/--
+A morphism of locally ringed spaces `f : X ⟶ Y` induces
+a local ring homomorphism from `Y.stalk (f x)` to `X.stalk x` for any `x : X`.
+-/
+def stalk_map {X Y : LocallyRingedSpace} (f : X ⟶ Y) (x : X) :
+  Y.stalk (f.1.1 x) ⟶ X.stalk x :=
+PresheafedSpace.stalk_map f.1 x
+
+instance {X Y : LocallyRingedSpace} (f : X ⟶ Y) (x : X) :
+  is_local_ring_hom (stalk_map f x) := f.2 x
+
+/-- The identity morphism on a locally ringed space. -/
+@[simps]
+def id (X : LocallyRingedSpace) : hom X X :=
+⟨𝟙 _, λ x, by { erw PresheafedSpace.stalk_map.id, apply is_local_ring_hom_id, }⟩
+
+instance (X : LocallyRingedSpace) : inhabited (hom X X) := ⟨id X⟩
+
+/-- Composition of morphisms of locally ringed spaces. -/
+@[simps]
+def comp {X Y Z : LocallyRingedSpace} (f : hom X Y) (g : hom Y Z) : hom X Z :=
+⟨f.val ≫ g.val, λ x,
+begin
+  erw PresheafedSpace.stalk_map.comp,
+  exact @is_local_ring_hom_comp _ _ _ _ _ _ _ _ (f.2 _) (g.2 _),
+end⟩
+
+/-- The category of locally ringed spaces. -/
+instance : category LocallyRingedSpace :=
+{ hom := hom,
+  id := id,
+  comp := λ X Y Z f g, comp f g,
+  comp_id' := by { intros, ext1, simp, },
+  id_comp' := by { intros, ext1, simp, },
+  assoc' := by { intros, ext1, simp, }, }.
```

Comment:
There's support for building full subcategories, but no existing API for categories where there's a condition on the morphisms too. I guess it would be useful, in order to automatically obtain the inclusion functor.

Reply:
I've added that functor now, as `forget_to_SheafedSpace`.